### PR TITLE
[FEAT]: daft sql

### PR DESCRIPTION
**Pull request by universalmind303**
Still TODO:

- [ ] python front-end
- [ ] add support for more functions
- [ ] group by
- [ ] more test cases
- [ ] at least 1 passing test for a tpch query

closes https://github.com/Eventual-Inc/Daft/issues/2468

### CODE REVIEW

**Comment by RCHowell** [comment]
I pushed some plumbing this morning, https://github.com/RCHowell/Daft/commit/1bba79908fac2a312425c17e2cc3628092d0153f

Would you prefer PRs to Daft targeting main or an sql feature branch — this way we can collaborate on the same branch. Thank you.

**Comment by universalmind303** [comment]
> I pushed some plumbing this morning, [RCHowell@1bba799](https://github.com/RCHowell/Daft/commit/1bba79908fac2a312425c17e2cc3628092d0153f)
> 
> Would you prefer PRs to Daft targeting main or an sql feature branch — this way we can collaborate on the same branch. Thank you.

I think as long as we label the public sql apis as `experimental`/`unstable`, targeting main is fine.

**Comment by RCHowell** [comment]
Ok, I've PR'd the initial structure. Thanks.

https://github.com/Eventual-Inc/Daft/pull/2559

File: src/daft-sql/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[dependencies]
+common-error = {version = "0.2.0-dev0", path = "../common/error"}
+daft-core = {version = "0.2.0-dev0", path = "../daft-core"}
+daft-dsl = {version = "0.2.0-dev0", path = "../daft-dsl"}
+daft-plan = {version = "0.2.0-dev0", path = "../daft-plan"}
```
Comment (jaychia) [inline src/daft-sql/Cargo.toml:5]: How come the versions have to be pinned like so?

File: src/daft-sql/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1041 @@
...
+        TableWithJoins, TimezoneInfo, UnaryOperator, Value, WildcardAdditionalOptions,
+    },
+    dialect::GenericDialect,
+    parser::{Parser, ParserOptions},
+    tokenizer::Tokenizer,
+};
+
+macro_rules! sql_bail {
+    ($($arg:tt)*) => {
+        return Err(DaftError::ValueError(format!($($arg)*)))
```
Comment (jaychia) [inline src/daft-sql/src/lib.rs:27]: Let's use Snafu so we can get nice errors from the crate?

**Review by jaychia — APPROVED** [review]
Seems like a good starting point!

File: src/daft-sql/Cargo.toml (hunk 1 above)
Comment (universalmind303) [reply src/daft-sql/Cargo.toml:5]: oh, i just ran `cargo add -p daft-sql <crate>` and it did this. The version doesn't have to be pinned unless you plan on releasing to crates.io

**Review by jaychia — APPROVED** [review]

**Pull request merged**